Feat/issue 154/configuring linting

jimbandrews opened the pull request:
## **Pull Request Template**

### **1. Targeted Issue** ###

*(The first step to a good solution is knowing the problem. Please link the issue here.)*

Issue #... or
Ticket #...

https://trello.com/c/UIf8r7Yr/154-add-additional-linting-plugins-to-improve-overall-code-consistency

### **2. Overview of Solution** ###

*(You've put in the hard work, now's your chance to tell us about it. What's your solution in a nutshell?)*

I've configured linting and auto-formatting to be performed whenever a contributor tries to make a commit. I've also added documentation detailing the linting plugins and how to setup pre-commit.


### **3. Tools Used** ###

*(Share the tools, libraries, or techniques you've used in your solution.)*

pre-commit, flake8, isort, black


### **4. Testing Strategy** ###

*(No code is complete without tests. Outline what you've tested and how.)*

1) Make a change to the codebase that is technically correct, but difficult to read.
2) Try to make a commit with a message that doesn't follow the schema in the docs.

You should see that the commit fails the isort, black, or flake8 steps, and fails the commit message step.
If isort or black fail, there will be new changes to commit.
In the end, the commit is cancelled.


### **5. Future Implications** ###

*(Let's anticipate potential effects of your changes. Are there any potential issues down the line?)*

Contributors will now have to follow the code style enforced by the linter and auto-formatters, leading to a more consistent codebase. If we want to require type hints in the future, we can use a flake8 plugin to include rules in the linter.


### **6. Screenshots** ###
<img width="711" alt="Screenshot 2023-10-25 at 1 48 06 PM" src="https://github.com/TeamProductionSystem/Momentors_Backend/assets/8958872/6d4c3cb6-54e0-49b2-aa3d-4238d6cad2cb">
<img width="711" alt="Screenshot 2023-10-25 at 1 48 42 PM" src="https://github.com/TeamProductionSystem/Momentors_Backend/assets/8958872/c300f839-e92b-4314-bdeb-90f8cc1baf46">

*(A picture is worth a thousand words. Include screenshots of your changes if relevant.)*


### **7. Code Reviewers** ###

*(Who else is in the loop? Mention your team or other contributors here.)*

@esparr @GitLukeW 


---

Before you go...

- [x ]  I have linked the relevant issue
- [x ]  I have provided an overview of my solution
- [x ]  I have shared the tools I've used
- [x ]  I have outlined my testing strategy
- [x ]  I have considered future implications of my changes
- [x ]  I have included relevant screenshots
- [x ]  I have kept my team in the loop

---

### Big high-fives for your incredible pull request! We're thrilled to have you as a contributor! Our team will dig into your proposed changes and give you some feedback or merge them if they're good to go. Your contributions mean the world to us!

jimbandrews committed:
added several flake8 plugins and black

jimbandrews committed:
removed flake8-isort; celery app is instantiated based on new ENVIRONMENT env variable

jimbandrews committed:
fixed docs on running celery server locally; added error handling for incorrect ENVIRONMENT value

jimbandrews committed:
fixing linting errors

jimbandrews committed:
added flake8-clean-block and fixed errors

jimbandrews committed:
added flake8-multiline-containers plugin and fixed errors

jimbandrews committed:
fixed multiline container in archive_session_list

jimbandrews committed:
fixed indentation

jimbandrews committed:
re-added black to pre-commit

jimbandrews committed:
corrected Environment Variables section in README; added flake8 plugins as dependencies in pre-commit config

jimbandrews committed:
fix(precommit): isort, black, flake8, and Branch Name Lint now only run during pre-commit hook

jimbandrews committed:
ci(linting): added flake8-isort plugin

jimbandrews committed:
docs(hooks): updating the Submitting Code section

jimbandrews committed:
docs(linting): added docs on what flake8 errors look like and each plugin's error code prefix

jimbandrews committed:
fix: commit-msg-linter no longer requires scope; removed autopep8

jimbandrews committed:
fix(readme): minor formatting JSONs in README

jimbandrews committed:
fix: Branch Name Lint should only run on push

jimbandrews committed:
docs: added example to branch name schema

jimbandrews committed:
docs: instructions on dealing with pre-commit during rebase

esparr commented on config/celery_settings.py:
```diff
@@ -1,23 +1,35 @@
+elif your_env == 'dev':
+    app = Celery('config')
+else:
+    raise ValueError('ENVIRONMENT in .env file should be either "dev" or "prod".')
```
@GitLukeW Would we need to have a staging env set up for the staging site or would we just use prod as well for it?

jimbandrews replied:
I would assume the staging environment would use the same app as prod, but that's just a guess.

GitLukeW replied:
I guess my question here is why are we changing this file?

jimbandrews replied:
I'm changing it because we want to remove all commented out code. The production code was important, so I created a new ENVIRONMENT env variable that we could check to see if it should be run. @esparr does make a good point, though. Is there anything that needs to run differently for staging?

esparr replied:
@GitLukeW Do you see any issues with this variable getting added to production? Seems like Staging would be set up like development for this one since we do not have a redis/celery setup on staging. 

One thing we will need to look out for is testing that celery is working once we get it on production (staging does not have celery/redis setup) I wonder if adding an integration test would be helpful to make sure this is working moving forward?

GitLukeW replied:
Sorry for the confusion. This can actually be taken out and just use the:

app = Celery('config')

esparr commented on scripts/commit-msg-lint.sh:
```diff
@@ -3,11 +3,11 @@
+    echo "Expected Format: <category(scope): description>"
     echo "Example: chore(test): setup test"
     exit 1
 fi
```
Is this lingering text or needed?

esparr commented on requirements.txt:
@GitLukeW I am wondering if bumping all of these dependencies will cause any issues?

jimbandrews replied:
These bumps are due to not specifying the version when the package was first added to the project. In the Pipfile, the star(*) indicates that pipenv should look for the newest version.

esparr commented on .flake8:
```diff
@@ -4,4 +4,7 @@ exclude =
     ./config/settings.py
     ./team_production_system/management/commands/add_superuser.py
-
+max-line-length = 80
```
There might be a conflict with this and black auto linting. I believe you have black set to max line length of 88.

jimbandrews replied:
The bugbear plugin allows for a 10% safety net, so the max-line-length is effectively 88. This is implemented by lines 8 and 9. I've added comments to address this.

esparr commented on pyproject.toml:
```diff
@@ -0,0 +1,9 @@
+[tool.black]
+line-length = 88
```
Check that this is consistent between black and flake8

jimbandrews replied:
See above comment on the .flake8 review

esparr commented on team_production_system/serializers/availability.py:
```diff
@@ -29,34 +35,28 @@ def create(self, validated_data):
         # Check if end time is between a start time and end time of
         # an existing availability
         overlapping_condition2 = Availability.objects.filter(
```
Instead of whitelisting conditions1, let's change it to condition_1 since condition passes the spellcheck

esparr commented on team_production_system/serializers/availability.py:
```diff
@@ -29,34 +35,28 @@ def create(self, validated_data):
+            mentor=mentor, start_time__lt=end_time, end_time__gt=end_time
         ).exists()
         # Check if start time is between the new start_time and new end_time
         overlapping_condition3 = Availability.objects.filter(
```
condition_3

esparr commented on team_production_system/serializers/availability.py:
```diff
@@ -29,34 +35,28 @@ def create(self, validated_data):
+            mentor=mentor, start_time__gte=start_time, start_time__lt=end_time
         ).exists()
         # Check if end time is between the new start_time and new end_time
         overlapping_condition4 = Availability.objects.filter(
```
condition_4

esparr commented on team_production_system/serializers/availability.py:
```diff
@@ -29,34 +35,28 @@ def create(self, validated_data):
         ).exists()
 
         availability_overlap = (
             overlapping_condition1
```
One comment for changing all these function condition arguments to be condition_x.

esparr commented on whitelist.txt:
```diff
@@ -0,0 +1,17 @@
+modelfields
+multiselectfield
+jitsi
+condition1
```
remove condition variations from whitelist.

esparr reviewed (changes requested):
This is looking really good. I am excited about more code consistency that is easy to run. I would like to talk more with @GitLukeW about the changes in more general terms and the environment setup. We might need to have some future tickets to deal more things in regards to environments. I left a few suggestions and a few comments for Luke to also look at.

jimbandrews committed:
removed commented out code

GitLukeW reviewed (changes requested):
Looks like it has some conflicts needing action prior to merging it.

jimbandrews committed:
added and configured isort to project; fixed imports on other files

jimbandrews committed:
added and configured isort; isort and black configs are now in pyproject.toml; fixed imports; added isort, black, and flake8 to pre-commit

jimbandrews committed:
added flake8-quotes plugin; fixed linting errors

jimbandrews committed:
docs: adding $ to shell commands

jimbandrews committed:
refactor: moved flake8 from packages to dev-packages in Pipfile; locked Pipfile; updated requirements.txt

jimbandrews committed:
chore: addressing PR comments

jimbandrews committed:
docs: auto-formatted and linted after rebase

GitLukeW commented on README.md:
```diff
@@ -267,11 +331,16 @@ Branch names should be in the following format:
 Commit messages should be in the following format:
 
-`<type>(<scope>): <description>`
+`<type>: <description>`
```
Why are we changing the format here?

jimbandrews replied:
The regex as it was didn't leave the scope as optional, but required it. I reached out about it in the general-backend discord channel. @esparr preferred removing it so the commit message wasn't clogged. What are your thoughts?

GitLukeW commented on README.md:
```diff
@@ -287,8 +356,6 @@ Commit messages should be in the following format:
 - chore - Miscellaneous changes, such as updating packages or bumping a version number
 - revert - Reverting a previous commit
 
```
This correlates to the question on 343.

GitLukeW commented on scripts/commit-msg-lint.sh:
```diff
@@ -3,11 +3,11 @@
 commit_msg=$(cat $commit_msg_file)
 
-regex="^(feat|fix|docs|style|refactor|perf|test|build|ci|chore|revert)\([a-z]+\): .+$"
+regex="^(feat|fix|docs|style|refactor|perf|test|build|ci|chore|revert): .+$"
```
Same question here. Why are we changing this?

jimbandrews replied:
This is changed as a result of removing the scope from commit message. My reply to your comment in the README expands on it.

(Not sure why I can't respond to the celery_settings.py comment, so I'm including it here)
We're changing the code in celery_settings.py because we want to remove all commented out code. The production code was important, so I created a new ENVIRONMENT env variable that we could check to see if it should be run. @esparr made a comment above wondering if we also need specific settings for the staging environment, apart from dev and prod.

GitLukeW commented on Pipfile:
```diff
@@ -20,17 +20,27 @@ django-imagekit = "*"
 django-multiselectfield = "*"
 boto3 = "*"
-sentry-sdk = {extras = ["django"], version = "*"}
+sentry-sdk = {extras = ["django"] }
```
Why is this changing?

jimbandrews replied:
I'm not sure. I'll run `pipenv install` again and see if that changes anything.

jimbandrews replied:
I've reverted it.

jimbandrews committed:
chore: reverting sentry-sdk installation

jimbandrews committed:
fix: commit script should now handle scope conditionally

jimbandrews committed:
fix(commit): testing commit script with scope

jimbandrews committed:
docs(commits): added optional scope back to documentation about how we lint the commit message

esparr commented on requirements.txt:
This file has been removed from github using .gitignore now. To remove it you will need to do the following steps: 
`git rm --cached requirements.txt`
`git commit -m "chore(gitignore): removes requirements.txt from cache"`

jimbandrews replied:
Done!

esparr commented on team_production_system/tests/unit_tests/serializers/test_availability_serializer.py:
I am getting this error when I run tests.

<img width="574" alt="Screen Shot 2023-10-31 at 10 24 01 AM" src="https://github.com/TeamProductionSystem/Momentors_Backend/assets/14296669/1dc61214-7d39-4612-89fd-c57ce0749ff6">

jimbandrews replied:
I found the error. Flake8 was mad about using the set constructor and a list literal. The expected keys are now created using a set literal. Also, I'm going to look into adding testing to the pre-commit hook.

esparr replied:
Amazing! Thanks for looking into it.

esparr reviewed (changes requested):
This is looking good! I found a couple things that need addressing before moving forward. This is a pretty big PR so we need to give it a good amount of looking over. Thanks for getting us through it.

jimbandrews committed:
added several flake8 plugins and black

removed flake8-isort; celery app is instantiated based on new ENVIRONMENT env variable

fixed docs on running celery server locally; added error handling for incorrect ENVIRONMENT value

removed commented out code

fixing linting errors

added and configured isort; isort and black configs are now in pyproject.toml; fixed imports; added isort, black, and flake8 to pre-commit

corrected Environment Variables section in README; added flake8 plugins as dependencies in pre-commit config

fix(precommit): isort, black, flake8, and Branch Name Lint now only run during pre-commit hook

docs(hooks): updating the Submitting Code section

docs(linting): added docs on what flake8 errors look like and each plugin's error code prefix

docs: added example to branch name schema

chore: addressing PR comments

docs: auto-formatted and linted after rebase

fix: commit script should now handle scope conditionally

docs(commits): added optional scope back to documentation about how we lint the commit message

chore(gitignore): removing requirements.txt from cache

fix(testing): test set created with set literal instead of set() constructor; added breakpoint to whitelist

chore(hooks): switching run tests to pre-push hook

chore(docs): updating Submitting Code section in README to include pre-push git hook

fix(rebasing): moved branch name linter to pre-push to prevent issues with rebasing

jimbandrews commented:
I've made another change to the git hooks. The Branch Linter now runs during the pre-push hook instead of pre-commit. The linter was failing while I tried to squash commits via rebasing when it was run during pre-commit.

esparr committed:
Merge branch 'feat/issue-154/configuring-linting' of https://github.com/jimbandrews/Team_Production_System_BE into feat/issue-154/configuring-linting

jimbandrews committed:
chore: removed 'production' code from celery_settings

jimbandrews committed:
Merge branch 'staging' into feat/issue-154/configuring-linting

jimbandrews committed:
Merge branch 'staging' into feat/issue-154/configuring-linting

jimbandrews committed:
fix: deleted Pipfile.lock and reinstalled

jimbandrews committed:
fix(docker): added 'localhost' to ALLOWED_HOSTS if DEBUG is true; fixes trying to call on locally-running container API

jimbandrews committed:
fix: removed deleted code that made it through merge/rebase

jimbandrews committed:
fix: removed $ from shell commands

GitLukeW reviewed (approved):
Thanks for the update! Lets get this thing merged!

esparr reviewed (approved):
Excited to get this in! Let's goooooooooooo!
